### PR TITLE
TST: stats: make `check_sample_var` two-sided

### LINKED ISSUE #2071: TST: stats: `check_sample_var` should be two-sided (Trac #1546)
_Original ticket http://projects.scipy.org/scipy/ticket/1546 on 2011-10-30 by @josef-pkt, assigned to unknown._

check_sample_var doesn't test anything

see gh-2070

### PR DESCRIPTION
**Pull request by mdhaber**
#### Note
CI failures are intentional; this is ready for review.

#### Reference issue
Closes gh-2071

#### What does this implement/fix?
`check_sample_var` is a unit test that runs for all continuous distributions (unless they are skipped):

```python3
def check_sample_var(sv, n, popvar):
    # two-sided chisquare test for sample variance equal to
    # hypothesized variance
    df = n-1
    chi2 = (n - 1)*sv/popvar
    pval = stats.distributions.chi2.sf(chi2, df) * 2
    npt.assert_(pval > 0.01, 'var fail, t, pval = %f, %f, v, sv=%f, %f' %
                (chi2, pval, popvar, sv))
```

`sv` is the sample variance of `n` observations from a distribution with population variance `popvar`; it is intended to test for unlikely deviations between the sample variance and population variance (similar to how a one-sample t-test checks for differences between sample mean and population mean). A reference for the test is [here](https://www.itl.nist.gov/div898/handbook/eda/section3/eda358.htm).

One big difference between this test and a t-test is that the [`chi2`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2.html) distribution has positive support (variance is always positive). Note that despite the comment `two-sided chisquare test for sample variance` and the multiplication by two, **the test above is one-sided**: `stats.distributions.chi2.sf(chi2, df)` is the probability under the null hypothesis of obtaining a sample variance _greater_ than that of the sample. If the sample variance is too small, then `pval` is large (e.g. `stats.distributions.chi2.sf(0, df)` is 1), and the assertion is trivially satisfied.

This PR currently corrects the test as suggested by @josef-pkt in gh-2071, but many distributions would fail the test. 
This is probably not due to unlikely sample variances but the fact that the distributions are non-normal. (The `chi2` distribution of the test statistic is derived under the assumption that the sample is drawn from a normal distribution.)

I don't think all these distributions are wrong, so I'd suggest we just remove the test as it seems too prone to false-positives.

### CODE REVIEW

**Comment by tupui** [comment]
Fine with me to remove this test.

**Comment by mdhaber** [comment]
@ev-br Does it sound reasonable to eliminate this test if it is prone to false positives given that the null distribution of the test statistic (`chi2`) is derived under the assumption of sample normality (which is obviously not true for non-normal distributions)?

**Comment by ev-br** [comment]
It does to me, yes. I'm not a statistician however, so maybe Chris or Josef could weigh in?

**Comment by mdhaber** [comment]
Just to follow up, this is what the null distribution of the test statistic looks like 500 samples from the normal distribution. Blue is the `chi2` pdf, orange is a histogram generated from `monte_carlo_test`.
image

Now orange is the null distribution for 500 samples from `gilbrat`:
image

It looks to me like our implementation of `var` and `rvs` for `gilbrat` are correct, but based on the actual (Monte Carlo) distribution of the test statistic, no wonder the test fails.

**Comment by mdhaber** [comment]
@josef-pkt This PR makes the change you suggested in gh-2071, but as you can see, there are many false positives (test failures for distributions that are OK). Can we just remove this test?

Note that we have plans to add many other tests using hypothesis to make sure that the various methods of distributions are consistent with one another; I'm just suggesting that the `check_sample_var` test is not great for this purpose.

**Comment by mdhaber** [comment]
What do you say, @josef-pkt? As you discovered, it's been broken for over ten years, so I don't think we'll miss it, and fixing it is going up bring up a bunch of false positives. Can we just remove it? Thanks for your thoughts!

**Comment by steppi** [comment]
For now I think it's fine to get rid of the test. I did a brief investigation though. It seemed to me like there should still be some kind of test so I decided to try a bootstrap to see what would happen. I wrote the following

```python
def check_sample_var(sample, popvar):
    res = stats.bootstrap(
        sample[np.newaxis, :],
        lambda x, axis: x.var(ddof=1, axis=axis),
        confidence_level=0.995,
    )
    conf = res.confidence_interval
    low, high = conf.low, conf.high
    assert low <= popvar <= high
```

which just generates a bootstrap confidence interval and checks if it contains the population variance. Using this test, there is only one failure for `laplace_asymmetric`. The default inverse transform sampling is used for this distribution. I reran the equivalent of this test for 1000 different samples of size 500 and there were only 20 failures out of the 1000 runs.

I wanted to test something out and rewrote `_rvs` as follows using the property that the asymmetric laplace distribution is the difference of exponentials.

```python
    def _rvs(self, kappa, size=None, random_state=None):
        A = random_state.exponential(1/kappa, size=size)
        B = random_state.exponential(kappa, size=size)
        return A - B
```

after this the test no longer fails. I tried rerunning 1000 iterations of the test and there were 19 failures. I checked some goodness of fit results and it seems that neither method is preferable over the other in terms of accuracy.

I'm not sure if there would be any interest in replacing `check_sample_var` with a bootstrap based test and increasing the sample size for this one failing test or something, but for now I think it's good to just get rid of the existing test.

**Comment by mdhaber** [comment]
> It seems the failure is partly a case of bad luck.

Why partly? I imagine that the speed of the RVS method is improved, but is the default RVS not accurate? I'd be happy to take a look at a PR upgrading `asymmetric_laplace` RVS, but it's not unreasonable that one test fails given the number of distributions we have.

Sure, the bootstrap idea seems like a suitable replacement. On my machine it takes only 60 ms, and I think it's faster than a Monte Carlo test would be (for most distributions) since it only requires one sample from the distribution.  I'll go ahead and replace it here.

File: scipy/stats/tests/test_continuous_basic.py [hunk 1]
```diff
@@ -493,10 +493,9 @@ def check_sample_var(sv, n, popvar):
     # two-sided chisquare test for sample variance equal to
     # hypothesized variance
     df = n-1
```
Comment (mdhaber) [inline scipy/stats/tests/test_continuous_basic.py:495]:
```suggestion
def check_sample_var(sample, popvar):
    res = stats.bootstrap(
        (sample,)
        lambda x, axis: x.var(ddof=1, axis=axis),
        confidence_level=0.995,
    )
    conf = res.confidence_interval
    low, high = conf.low, conf.high
    assert low <= popvar <= high
```

File: scipy/stats/tests/test_continuous_basic.py [hunk 2]
```diff
@@ -493,10 +493,9 @@ def check_sample_var(sv, n, popvar):
...
-    chi2 = (n - 1)*sv/popvar
-    pval = stats.distributions.chi2.sf(chi2, df) * 2
-    npt.assert_(pval > 0.01, 'var fail, t, pval = %f, %f, v, sv=%f, %f' %
-                (chi2, pval, popvar, sv))
+    statistic = df * sv / popvar
+    pval = min(stats.chi2.cdf(statistic, df), stats.chi2.sf(statistic, df))
+    assert pval > 0.005
```
Comment (mdhaber) [inline scipy/stats/tests/test_continuous_basic.py:498]:
```suggestion
```

**Comment by steppi** [comment]
> Thanks for taking a look @steppi!
> 
> > It seems the failure is partly a case of bad luck.
> 
> Why partly? I'd be happy to take a look at a PR upgrading `asymmetric_laplace` RVS, but it's not wrong, right? I think it's OK that one test fails given the confidence level and the number of distributions we have.
> 
> Sure, the bootstrap idea seems like a good replacement. On my machine it takes only 60 ms, and I think it's faster than a Monte Carlo test would be (for most distributions) since it only requires one sample from the distribution. I'll go ahead and replace the test here.

Sorry, I was a little unclear. The implementation is absolutely correct and I'm not sure it can be improved on. I said partly, because from my experiments it seems like the sample variance converges to the population variance slowly for this distribution compared to the others. For others I've checked the failure rate is lower, in line with the 0.995 confidence level. Partly because the nature of the distribution make it seem like empirically we get 0.980 confidence level for the test instead of 0.995 :)

**Comment by mdhaber** [comment]
> If the test fails, it could either be that stats does not compute the variance correctly or that rvs does not sample from the desired distribution

Yes, or - as it is now - that the test is not appropriate for the data OR that it is a reasonable random false positive. But yes, what you have written is the _intent_ of the test.

> For the latter, we should have stronger tests (relying on the KS test if I remember correctly).

Yes, we already do a KS test that checks consistency between the RVS and the distribution:
https://github.com/scipy/scipy/blob/3292e2529d9e6311d65f06b9af4f5ec067fb9143/scipy/stats/tests/test_continuous_basic.py#L595-L611

> So unless we believe the additional check is needed, we can also consider removing it. It probably also depends on how time-consuming `stats.bootstrap` is.

Yup, and that's what I proposed, too, but since some people think it's a good idea, I don't see harm including it. I found above that on my machine it takes only about 60 ms, and that should be independent of the distribution.

**Comment by josef-pkt** [comment]
sorry, I didn't see this PR.

just for checking:
Can you run the unit tests with a much larger sample size and see if more unit tests pass.

When I initially wrote the random sample based unit tests, then the sample size was considerably larger.
The sample size was reduced over time to reduce the time that the unit tests take.

convergence to normality for test statistics can be very slow for some higher moment cases. Variance tests are pretty bad in small sample if distribution is far from normal.

In general, those random sample based unit test were the easiest to check a large number of distributions without going into distribution specific details.
If there are alternative tests that check statistics like variance, then those random sample tests can be dropped.
e.g. we have unit test to check `expect` with `var`, AVAIR.

(Another possibility might be to use improved tests instead of the standard chisquare test, e.g. adding a correction for kurtosis. But I don't have those tests yet in statsmodels, they are just on my long term wish list. )

**Comment by josef-pkt** [comment]
Note, the initial commit in this PR does not have the correct p-value computation, it misses a `*2`

```
    statistic = df * sv / popvar
    pval = min(stats.chi2.cdf(statistic, df), stats.chi2.sf(statistic, df))
    assert pval > 0.005
```

should have
`pval = 2 * min(stats.chi2.cdf(statistic, df), stats.chi2.sf(statistic, df))`

I got confused because  adding the check on the lower tail should fail if the empirical variance is too small which I would have expected with short tailed distributions.

But the failing gilbrat/lognormal case has heavy tails (large kurtosis) and it should fail with empirical variance too large if there are heavy tail outliers.
IIUC what happened here.

**Comment by josef-pkt** [comment]
Aside: I never heard of the gilbrat distribution.

However, name seems to be misspelled with an extra letter `l`
https://mathworld.wolfram.com/GibratsDistribution.html

**Comment by mdhaber** [comment]
> the initial commit in this PR does not have the correct p-value computation

It compares it against a value that is half of the original value, so it's the same, effectively. We get to choose that value, after all.
I agree that if we were returning a two-sided p-value we would multiply by two or follow another convention to get a two-sided p-value from an asymmetric distribution, but here I think it's understood that it's the probability in one tail.

**Comment by josef-pkt** [comment]
my suggestion would be to drop the rvs sample based unit test for var and check that `var` correctness is covered by unit test using `expect`, i.e. numerical integration.

The bootstrap version looks fine but might not contribute much if expect checks var.

**Comment by mdhaber** [comment]
14 distributions fail the original test with `sn=1000`, which was the original sample size from 14 years ago. 15 fail with `sn=10000`.

**Comment by josef-pkt** [comment]
> 14 distributions fail the original test with sn=1000, which was the original sample size from 14 years ago. 15 fail with sn=10000

thanks for checking
sounds pretty bad, I would have expected fewer failures
However, it confirms the bad reputation of variance tests for non-normal distributions.

**Comment by steppi** [comment]
I thought about this a bit, and I actually think that the bootstrap test can contribute something that the KS statistic doesn't, since KS isn't sensitive to behavior in the tails. I came up with an example to illustrate.

```python
import numpy as np
import scipy.stats as stats


random_state = np.random.default_rng(1234)
stats.pareto.random_state = random_state

# Generates from pareto distribution and clips at x=5.
# Sample size same as KS tests in test_distributions.py
x = stats.pareto.rvs(3, size=1500)
x = np.where(x >= 5, 5, x)

print(f"Expected variance is {stats.pareto.var(3)}, got {x.var(ddof=1)}")
# Expected variance is 0.75, got 0.3912855049688287

print(stats.ks_1samp(x, lambda x: stats.pareto.cdf(x, 3)))
# KstestResult(statistic=0.013377694524758232, pvalue=0.9478381367985416)

# Sample size the same as check_var tests in test_continuous_basic.py
x = stats.pareto.rvs(3, size=500)
x = np.where(x >= 5, 5, x)

print(f"Expected variance is {stats.pareto.var(3)}, got {x.var(ddof=1)}")
# Expected variance is 0.75, got 0.36957201201232076

result = stats.bootstrap((x, ), lambda x, axis: x.var(ddof=1, axis=axis))
conf = result.confidence_interval
print(f"Confidence interval ({conf.low}, {conf.high})) does not contain 0.75.")
# Confidence interval (0.288385871630806, 0.49552553841346075)) does not contain 0.75.
```

KS is struggles to detect the clipping here, but the variance test picks it up easily. Perhaps others will find the example contrived, but this kind of clipping seems like something that could happen in the wild. I've convinced myself to be +1 for including the test, although perhaps there are other tests that would be even better.

**Comment by mdhaber** [comment]
Thanks @steppi. I'm happy with keeping it. Sure, there might be better tests, and those can come as enhancements in time, but I think it is enough to close this very old issue : )

**Comment by mdhaber** [comment]
>  check that var correctness is covered by unit test using expect, i.e. numerical integration.

@josef-pkt Yes, this is done.

https://github.com/scipy/scipy/blob/6f3c7be0961c24b9086ed8e5f827e030b313fd5d/scipy/stats/tests/test_continuous_basic.py#L275-L307

**Review by steppi — APPROVED** [review]
I think this looks good and am +1 for including it. I edited in some lines to the snippet I wrote earlier that show that `check_sample_mean` is unable to detect the clipping in that example.

**Comment by mdhaber** [comment]
Thanks @steppi! What do you think about merging on Monday if there are no more comments this weekend?

**Comment by steppi** [comment]
> Thanks @steppi! What do you think about merging on Monday if there are no more comments this weekend?

Sounds good to me.

**Comment by steppi** [comment]
Will merge at around 5pm EDT (UTC-04:00) Monday if there are no objections.

**Comment by chrisb83** [comment]
Given that we have `test_moments` (test moments computed by stats method against results of a numerical integration) and `check_distribution_rvs` (KS test), I would prefer to remove this check altogether. 

> This and check_sample_mean are just additional lines of defense since individually not rejecting the null hypothesis is not really much evidence for the null hypothesis.

I agree it is an additional check, though a very weak one. The problem that you describe (accepting the null hypothesis if the alternative is true) is related to the power of a test: the KS test has known weaknesses in this regard. one could use stronger tests like CvM (if I recall correctly, its power is greater than KS...), it is just that `check_sample_mean` does not add a lot of value (in my opinion :)

If you prefer to merge, I am fine, ~6 seconds more (assuming ~100 distributions) is not an issue

**Comment by chrisb83** [comment]
> I thought about this a bit, and I actually think that the bootstrap test can contribute something that the KS statistic doesn't, since KS isn't sensitive to behavior in the tails. I came up with an example to illustrate.

@steppi I did not pay attention to this example... It shows a relevant use case, so please go ahead and merge as planned :)

**Comment by mdhaber** [comment]
Thanks @steppi @chrisb83 @josef-pkt!

**Pull request merged**